scenes/test_menu.txt

```diff
@@ -0,0 +1,173 @@
+*set cslib_catch_choice true
+
+*comment :::::: SIMPLE tests ::::::
+*gosub test_start "simple - single item"
+*set cslib_choice_select "1"
+*gosub_scene cslib_menu build_simple "Hello"
+*gosub test_assert_equal 1 cslib_choice_count
+*gosub test_assert_equal 1 cslib_ret
+*gosub test_assert_equal "Hello" cslib_choice_1
+*gosub test_finish
+
+*gosub test_start "page-count invalid"
+*temp result_expected "Bug: cslib_menu._render_choice: this option should not be selectable"
+*set cslib_choice_select "1"
+*set cslib_catch_bug true
+*gosub_scene cslib_menu build_simple 1 2 3 4 5 6
+*set cslib_catch_bug false
+*gosub test_finish
+
+*gosub test_start "simple - two items (with cancel)"
+*set cslib_choice_select "3"
+*gosub_scene cslib_menu build_simple_with_cancel "Hello" "Goodbye"
+*gosub test_assert_equal 3 cslib_choice_count
+*gosub test_assert_equal 0 cslib_ret
+*gosub test_assert_equal "Hello" cslib_choice_1
+*gosub test_assert_equal "Goodbye" cslib_choice_2
+*gosub test_assert_equal "[return]" cslib_choice_3
+*gosub test_finish
+
+*gosub test_start "simple - single-page (page 1, option 1)"
+*set cslib_choice_select "3"
+*gosub_scene cslib_menu build_simple 1 2 3 4 5 6 7 8 9 10 11
+*gosub test_assert_equal 7 cslib_choice_count
+*gosub test_assert_equal 1 cslib_ret
+*gosub test_assert_equal "Page (1 / 3)" cslib_choice_1
+*gosub test_assert_equal "[i]Next -->[/i]" cslib_choice_2
+*gosub test_assert_equal "1" cslib_choice_3
+*gosub test_assert_equal "2" cslib_choice_4
+*gosub test_assert_equal "3" cslib_choice_5
+*gosub test_assert_equal "4" cslib_choice_6
+*gosub test_assert_equal "5" cslib_choice_7
+*gosub test_finish
+
+*gosub test_start "simple - multi-page (navigation)"
+*comment Select 'Next' then 'Next' then 'Previous', then '8' on Page 2
+*set cslib_choice_select "2,2,2,5"
+*gosub_scene cslib_menu build_simple 1 2 3 4 5 6 7 8 9 10 11
+*gosub test_assert_equal 8 cslib_choice_count
+*gosub test_assert_equal 8 cslib_ret
+*gosub test_assert_equal "Page (2 / 3)" cslib_choice_1
+*comment due to 'smart' navigation, previous will be at the top
+*gosub test_assert_equal "[i]<-- Previous[/i]" cslib_choice_2
+*gosub test_assert_equal "6" cslib_choice_3
+*gosub test_assert_equal "7" cslib_choice_4
+*gosub test_assert_equal "8" cslib_choice_5
+*gosub test_assert_equal "9" cslib_choice_6
+*gosub test_assert_equal "10" cslib_choice_7
+*gosub test_assert_equal "[i]Next -->[/i]" cslib_choice_8
+*gosub test_finish
+
+*gosub test_start "simple - multi-page (page 3, option 1)"
+*set cslib_choice_select "2,2,3"
+*gosub_scene cslib_menu build_simple 1 2 3 4 5 6 7 8 9 10 11
+*gosub test_assert_equal 3 cslib_choice_count
+*gosub test_assert_equal 11 cslib_ret
+*gosub test_assert_equal "Page (3 / 3)" cslib_choice_1
+*gosub test_assert_equal "[i]<-- Previous[/i]" cslib_choice_2
+*gosub test_assert_equal "11" cslib_choice_3
+*gosub test_finish
+
+
+*comment :::::: ARRAY tests ::::::
+*comment array - init test data
+*set test_menu_opt_count 7
```

Comment:
I won't stop this, but the filters are adding a layer of complexity - we don't have a strong design for 'objects' yet (for instance, I would feel this more natural: menu_opt[2]["visible"] than menu_opt["visible"][2])

Reply:
If we had an array module that allowed you to filter values _before_ passing them to cslib_menu, then I might agree, but as it stands there is no other way to present only a subset of options -- which is a crucial feature.

We even use it in library testing framework itself: https://github.com/ChoicescriptIDE/cslib/blob/9e69e00e65399c239ec5f9cb023fb7abc9b0b1ee/scenes/_test_entry.txt#L61 -- To only display test logs with content.

Removing this would make cslib_menu useless outside of situations where you knew every single option was selectable (which I imagine is actually the less frequent of use-cases).

--

The whole thing is meant to be a "beta" right? So let's see what people think? Better for people to say they don't like it and remove it, then never get feedback on it at all.

Reply:
Agreed